packages/table/src/TableContext.tsx

```diff
@@ -0,0 +1,228 @@
+import React, { createContext, useContext, useMemo } from 'react';
+
+const Types = {
+  SelectableTable: 'SELECTABLE_TABLE',
+  ToggleHeaderCheckedState: 'TOGGLE_HEADER_CHECKED',
+  ToggleHeaderIndeterminate: 'TOGGLE_HEADER_INDETERMINATE',
+  SetColumnInfo: 'SET_COLUMN_INFO',
+  SortTableData: 'SORT_TABLE_DATA',
+  ToggleIndividualChecked: 'TOGGLE_INDIVIDUAL_CHECKED',
+  SetRowCheckedState: 'SET_ROW_CHECKED_STATE',
+  SetHasNestedRows: 'SET_HAS_NESTED_ROWS',
+  SetHasRowSpan: 'SET_HAS_ROW_SPAN',
+} as const;
+
+type Types = typeof Types[keyof typeof Types];
+
+export { Types };
+
+interface ActionPayload {
+  [Types.SelectableTable]: boolean;
+  [Types.ToggleHeaderCheckedState]: boolean | undefined;
+  [Types.ToggleHeaderIndeterminate]: boolean;
+  [Types.SetColumnInfo]: {
+    dataType?: DataType;
+    index: number;
+  };
+  [Types.SetHasRowSpan]: boolean;
+  [Types.SetHasNestedRows]: boolean;
+  [Types.SortTableData]: {
+    columnId: number;
+    key: string;
+    data: Array<unknown>;
+  };
+  [Types.ToggleIndividualChecked]: {
+    index: number;
+    checked?: boolean;
+  };
+  [Types.SetRowCheckedState]: Array<boolean | undefined>;
+}
+
+type ActionMap<A extends Record<string, any>> = {
+  [Key in keyof A]: A[Key] extends undefined
+    ? {
+        type: Key;
+      }
+    : {
+        type: Key;
+        payload: A[Key];
+      };
+};
+
+type Action = ActionMap<ActionPayload>[keyof ActionMap<ActionPayload>];
+
+type Dispatch = (action: Action) => void;
+
+interface Sort {
+  columnId?: number;
+  direction: 'asc' | 'desc';
+  key?: string;
+}
+
+const DataType = {
+  Number: 'Number',
+  Quantity: 'quantity',
+  Weight: 'weight',
+  ZipCode: 'zipCode',
+  String: 'string',
+  Date: 'date',
+} as const;
+
+type DataType = typeof DataType[keyof typeof DataType];
+
+export { DataType };
+
+export interface State {
+  sort?: Sort;
+  data: Array<any>;
+  columnInfo?: {
+    [k in number]: { dataType?: DataType };
+  };
+  selectable?: boolean;
+  headerCheckState?: boolean;
+  headerIndeterminate?: boolean;
+  rowCheckedState?: Record<number, boolean | undefined>;
```

Comment:
Do you think we could have this and the two header state's be simply booleans and not also undefined? It'd be nice if we didn't have to worry about the undefined case and simply rely on `true || false`.

Reply:
👍